File: src/backend/models/AuditLog.js
/**
 * Audit Log Model
 *
 * Records all significant system events for compliance, security monitoring,
 * and debugging purposes. Supports structured logging with full context.
 *
 * @module backend/models/AuditLog
 */

// -----------------------------------------------------------------------------
// Constants
// -----------------------------------------------------------------------------

/**
 * Audit event types categorized by domain
 */
const AUDIT_EVENT_TYPES = {
    // Authentication events
    AUTH: {
        LOGIN_SUCCESS: 'auth.login.success',
        LOGIN_FAILURE: 'auth.login.failure',
        LOGOUT: 'auth.logout',
        PASSWORD_CHANGE: 'auth.password.change',
        PASSWORD_RESET_REQUEST: 'auth.password.reset_request',
        PASSWORD_RESET_COMPLETE: 'auth.password.reset_complete',
        MFA_ENROLL: 'auth.mfa.enroll',
        MFA_CHALLENGE_SUCCESS: 'auth.mfa.success',
        MFA_CHALLENGE_FAILURE: 'auth.mfa.failure',
        TOKEN_REFRESH: 'auth.token.refresh',
        SESSION_EXPIRED: 'auth.session.expired',
    },

    // User management events
    USER: {
        CREATE: 'user.create',
        UPDATE: 'user.update',
        DELETE: 'user.delete',
        BLOCK: 'user.block',
        UNBLOCK: 'user.unblock',
        ROLE_ASSIGN: 'user.role.assign',
        ROLE_REMOVE: 'user.role.remove',
        INVITE: 'user.invite',
        ACTIVATE: 'user.activate',
    },

    // Team events
    TEAM: {
        CREATE: 'team.create',
        UPDATE: 'team.update',
        DELETE: 'team.delete',
        ARCHIVE: 'team.archive',
        MEMBER_ADD: 'team.member.add',
        MEMBER_REMOVE: 'team.member.remove',
        MEMBER_ROLE_CHANGE: 'team.member.role_change',
    },

    // Access events
    ACCESS: {
        PERMISSION_GRANTED: 'access.permission.granted',
        PERMISSION_DENIED: 'access.permission.denied',
        RESOURCE_ACCESS: 'access.resource',
        API_CALL: 'access.api',
    },

    // System events
    SYSTEM: {
        CONFIG_CHANGE: 'system.config.change',
        MAINTENANCE_START: 'system.maintenance.start',
        MAINTENANCE_END: 'system.maintenance.end',
        ERROR: 'system.error',
        SECURITY_ALERT: 'system.security.alert',
    },

    // Compliance events
    COMPLIANCE: {
        REPORT_GENERATE: 'compliance.report.generate',
        REPORT_DOWNLOAD: 'compliance.report.download',
        DATA_EXPORT: 'compliance.data.export',
        CONSENT_UPDATE: 'compliance.consent.update',
    },
};

/**
 * Severity levels for audit events
 */
const SEVERITY_LEVELS = {
    DEBUG: 'debug',
    INFO: 'info',
    NOTICE: 'notice',
    WARNING: 'warning',
    ERROR: 'error',
    CRITICAL: 'critical',
    ALERT: 'alert',
};

/**
 * Outcome status for audit events
 */
const OUTCOME_STATUS = {
    SUCCESS: 'success',
    FAILURE: 'failure',
    UNKNOWN: 'unknown',
};

// -----------------------------------------------------------------------------
// Model Schema Definition
// -----------------------------------------------------------------------------

/**
 * Audit log schema definition
 */
const AuditLogSchema = {
    tableName: 'audit_logs',
    timestamps: true,

    fields: {
        id: {
            type: 'uuid',
            primaryKey: true,
            defaultValue: 'uuid_generate_v4()',
        },
        eventType: {
            type: 'string',
            length: 100,
            allowNull: false,
            index: true,
            comment: 'Type of audit event (e.g., user.create, auth.login)',
        },
        eventCategory: {
            type: 'string',
            length: 50,
            allowNull: false,
            index: true,
            comment: 'Category grouping (auth, user, team, access, system)',
        },
        severity: {
            type: 'enum',
            values: Object.values(SEVERITY_LEVELS),
            defaultValue: SEVERITY_LEVELS.INFO,
        },
        outcome: {
            type: 'enum',
            values: Object.values(OUTCOME_STATUS),
            defaultValue: OUTCOME_STATUS.SUCCESS,
        },
        actorId: {
            type: 'string',
            length: 255,
            allowNull: true,
            index: true,
            comment: 'User ID of the actor who performed the action',
        },
        actorType: {
            type: 'string',
            length: 50,
            defaultValue: 'user',
            comment: 'Type of actor (user, system, api)',
        },
        actorEmail: {
            type: 'string',
            length: 255,
            allowNull: true,
        },
        actorIp: {
            type: 'string',
            length: 45,
            allowNull: true,
            comment: 'IP address of the actor',
        },
        targetType: {
            type: 'string',
            length: 50,
            allowNull: true,
            comment: 'Type of target entity (user, team, resource)',
        },
        targetId: {
            type: 'string',
            length: 255,
            allowNull: true,
            index: true,
            comment: 'ID of the target entity',
        },
        targetName: {
            type: 'string',
            length: 255,
            allowNull: true,
        },
        description: {
            type: 'text',
            allowNull: true,
            comment: 'Human-readable description of the event',
        },
        changes: {
            type: 'json',
            defaultValue: null,
            comment: 'JSON object containing before/after values for changes',
        },
        metadata: {
            type: 'json',
            defaultValue: {},
            comment: 'Additional context and metadata',
        },
        requestId: {
            type: 'string',
            length: 100,
            allowNull: true,
            index: true,
            comment: 'Correlation ID for request tracing',
        },
        userAgent: {
            type: 'text',
            allowNull: true,
        },
        location: {
            type: 'json',
            defaultValue: null,
            comment: 'Geolocation data if available',
        },
        timestamp: {
            type: 'datetime',
            defaultValue: 'NOW()',
            index: true,
        },
        createdAt: {
            type: 'datetime',
            defaultValue: 'NOW()',
        },
    },

    indexes: [
        { fields: ['eventType'] },
        { fields: ['eventCategory'] },
        { fields: ['actorId'] },
        { fields: ['targetId'] },
        { fields: ['timestamp'] },
        { fields: ['requestId'] },
        { fields: ['severity'] },
        { fields: ['outcome'] },
        // Composite indexes for common queries
        { fields: ['actorId', 'timestamp'] },
        { fields: ['eventCategory', 'timestamp'] },
        { fields: ['targetType', 'targetId', 'timestamp'] },
    ],
};

// -----------------------------------------------------------------------------
// Audit Log Model Class
// -----------------------------------------------------------------------------

/**
 * Audit log model with business logic
 */
class AuditLog {
    /**
     * Create a new AuditLog instance
     *
     * @param {Object} data - Audit log data
     */
    constructor(data = {}) {
        this.id = data.id || null;
        this.eventType = data.eventType || null;
        this.eventCategory = data.eventCategory || this._extractCategory(data.eventType);
        this.severity = data.severity || SEVERITY_LEVELS.INFO;
        this.outcome = data.outcome || OUTCOME_STATUS.SUCCESS;

        // Actor information
        this.actorId = data.actorId || null;
        this.actorType = data.actorType || 'user';
        this.actorEmail = data.actorEmail || null;
        this.actorIp = data.actorIp || null;

        // Target information
        this.targetType = data.targetType || null;
        this.targetId = data.targetId || null;
        this.targetName = data.targetName || null;

        // Event details
        this.description = data.description || null;
        this.changes = data.changes || null;
        this.metadata = data.metadata || {};

        // Request context
        this.requestId = data.requestId || null;
        this.userAgent = data.userAgent || null;
        this.location = data.location || null;

        // Timestamps
        this.timestamp = data.timestamp || new Date();
        this.createdAt = data.createdAt || new Date();
    }

    /**
     * Extract category from event type
     * @private
     */
    _extractCategory(eventType) {
        if (!eventType) return null;
        const parts = eventType.split('.');
        return parts[0] || null;
    }

    /**
     * Validate audit log data
     *
     * @returns {Object} Validation result
     */
    validate() {
        const errors = [];

        if (!this.eventType) {
            errors.push({ field: 'eventType', message: 'Event type is required' });
        }

        const validSeverities = Object.values(SEVERITY_LEVELS);
        if (!validSeverities.includes(this.severity)) {
            errors.push({ field: 'severity', message: `Severity must be one of: ${validSeverities.join(', ')}` });
        }

        const validOutcomes = Object.values(OUTCOME_STATUS);
        if (!validOutcomes.includes(this.outcome)) {
            errors.push({ field: 'outcome', message: `Outcome must be one of: ${validOutcomes.join(', ')}` });
        }

        return {
            isValid: errors.length === 0,
            errors,
        };
    }

    /**
     * Convert to plain object
     *
     * @returns {Object} Plain object representation
     */
    toJSON() {
        return {
            id: this.id,
            eventType: this.eventType,
            eventCategory: this.eventCategory,
            severity: this.severity,
            outcome: this.outcome,
            actor: {
                id: this.actorId,
                type: this.actorType,
                email: this.actorEmail,
                ip: this.actorIp,
            },
            target: {
                type: this.targetType,
                id: this.targetId,
                name: this.targetName,
            },
            description: this.description,
            changes: this.changes,
            metadata: this.metadata,
            requestId: this.requestId,
            userAgent: this.userAgent,
            location: this.location,
            timestamp: this.timestamp,
            createdAt: this.createdAt,
        };
    }

    /**
     * Convert to storage format
     *
     * @returns {Object} Flat object for database storage
     */
    toStorageFormat() {
        return {
            id: this.id,
            eventType: this.eventType,
            eventCategory: this.eventCategory,
            severity: this.severity,
            outcome: this.outcome,
            actorId: this.actorId,
            actorType: this.actorType,
            actorEmail: this.actorEmail,
            actorIp: this.actorIp,
            targetType: this.targetType,
            targetId: this.targetId,
            targetName: this.targetName,
            description: this.description,
            changes: this.changes,
            metadata: this.metadata,
            requestId: this.requestId,
            userAgent: this.userAgent,
            location: this.location,
            timestamp: this.timestamp,
            createdAt: this.createdAt,
        };
    }

    /**
     * Check if this is a security-related event
     *
     * @returns {boolean}
     */
    isSecurityEvent() {
        const securityCategories = ['auth', 'access'];
        const securitySeverities = [SEVERITY_LEVELS.WARNING, SEVERITY_LEVELS.ERROR, SEVERITY_LEVELS.CRITICAL, SEVERITY_LEVELS.ALERT];

        return securityCategories.includes(this.eventCategory) ||
               securitySeverities.includes(this.severity);
    }

    /**
     * Check if this is a failed event
     *
     * @returns {boolean}
     */
    isFailed() {
        return this.outcome === OUTCOME_STATUS.FAILURE;
    }

    /**
     * Check if this is a high-severity event
     *
     * @returns {boolean}
     */
    isHighSeverity() {
        return [SEVERITY_LEVELS.ERROR, SEVERITY_LEVELS.CRITICAL, SEVERITY_LEVELS.ALERT].includes(this.severity);
    }
}

// -----------------------------------------------------------------------------
// Audit Log Builder
// -----------------------------------------------------------------------------

/**
 * Builder pattern for creating audit logs
 */
class AuditLogBuilder {
    constructor() {
        this.data = {};
    }

    /**
     * Set event type
     *
     * @param {string} eventType - Event type
     * @returns {AuditLogBuilder}
     */
    event(eventType) {
        this.data.eventType = eventType;
        this.data.eventCategory = eventType.split('.')[0];
        return this;
    }

    /**
     * Set severity level
     *
     * @param {string} severity - Severity level
     * @returns {AuditLogBuilder}
     */
    severity(severity) {
        this.data.severity = severity;
        return this;
    }

    /**
     * Set outcome status
     *
     * @param {string} outcome - Outcome status
     * @returns {AuditLogBuilder}
     */
    outcome(outcome) {
        this.data.outcome = outcome;
        return this;
    }

    /**
     * Mark as success
     *
     * @returns {AuditLogBuilder}
     */
    success() {
        this.data.outcome = OUTCOME_STATUS.SUCCESS;
        return this;
    }

    /**
     * Mark as failure
     *
     * @returns {AuditLogBuilder}
     */
    failure() {
        this.data.outcome = OUTCOME_STATUS.FAILURE;
        return this;
    }

    /**
     * Set actor information
     *
     * @param {Object} actor - Actor data
     * @returns {AuditLogBuilder}
     */
    actor(actor) {
        this.data.actorId = actor.id || actor.sub;
        this.data.actorType = actor.type || 'user';
        this.data.actorEmail = actor.email;
        this.data.actorIp = actor.ip;
        return this;
    }

    /**
     * Set actor from request
     *
     * @param {Object} req - Express request object
     * @returns {AuditLogBuilder}
     */
    fromRequest(req) {
        if (req.user) {
            this.data.actorId = req.user.id;
            this.data.actorEmail = req.user.email;
            this.data.actorType = 'user';
        }
        this.data.actorIp = req.ip || req.connection?.remoteAddress;
        this.data.userAgent = req.get('User-Agent');
        this.data.requestId = req.id || req.get('X-Request-ID');
        return this;
    }

    /**
     * Set target information
     *
     * @param {string} type - Target type
     * @param {string} id - Target ID
     * @param {string} name - Target name
     * @returns {AuditLogBuilder}
     */
    target(type, id, name = null) {
        this.data.targetType = type;
        this.data.targetId = id;
        this.data.targetName = name;
        return this;
    }

    /**
     * Set description
     *
     * @param {string} description - Event description
     * @returns {AuditLogBuilder}
     */
    describe(description) {
        this.data.description = description;
        return this;
    }

    /**
     * Set changes (before/after values)
     *
     * @param {Object} before - Previous values
     * @param {Object} after - New values
     * @returns {AuditLogBuilder}
     */
    changes(before, after) {
        this.data.changes = { before, after };
        return this;
    }

    /**
     * Add metadata
     *
     * @param {Object} metadata - Additional metadata
     * @returns {AuditLogBuilder}
     */
    meta(metadata) {
        this.data.metadata = { ...this.data.metadata, ...metadata };
        return this;
    }

    /**
     * Set location data
     *
     * @param {Object} location - Geolocation data
     * @returns {AuditLogBuilder}
     */
    location(location) {
        this.data.location = location;
        return this;
    }

    /**
     * Build the audit log instance
     *
     * @returns {AuditLog}
     */
    build() {
        return new AuditLog(this.data);
    }
}

// -----------------------------------------------------------------------------
// Repository Implementation
// -----------------------------------------------------------------------------

/**
 * Audit log repository for data access
 */
class AuditLogRepository {
    constructor(db) {
        this.db = db;
        this.table = AuditLogSchema.tableName;
    }

    /**
     * Create a new audit log entry
     *
     * @param {AuditLog|Object} log - Audit log data
     * @returns {Promise<AuditLog>}
     */
    async create(log) {
        const logData = log instanceof AuditLog ? log.toStorageFormat() : log;
        const logInstance = new AuditLog(logData);

        const validation = logInstance.validate();
        if (!validation.isValid) {
            const error = new Error('Validation failed');
            error.errors = validation.errors;
            throw error;
        }

        const result = await this.db.insert({
            table: this.table,
            data: logInstance.toStorageFormat(),
        });

        return new AuditLog(result);
    }

    /**
     * Find audit log by ID
     *
     * @param {string} id - Log ID
     * @returns {Promise<AuditLog|null>}
     */
    async findById(id) {
        const result = await this.db.findOne({
            table: this.table,
            where: { id },
        });

        return result ? new AuditLog(result) : null;
    }

    /**
     * Query audit logs with filtering
     *
     * @param {Object} options - Query options
     * @returns {Promise<Object>}
     */
    async query(options = {}) {
        const {
            page = 0,
            perPage = 50,
            eventType,
            eventCategory,
            severity,
            outcome,
            actorId,
            targetType,
            targetId,
            startDate,
            endDate,
            search,
            sortBy = 'timestamp',
            sortOrder = 'DESC',
        } = options;

        const query = {
            table: this.table,
            where: {},
            limit: perPage,
            offset: page * perPage,
            orderBy: [[sortBy, sortOrder]],
        };

        // Apply filters
        if (eventType) query.where.eventType = eventType;
        if (eventCategory) query.where.eventCategory = eventCategory;
        if (severity) query.where.severity = severity;
        if (outcome) query.where.outcome = outcome;
        if (actorId) query.where.actorId = actorId;
        if (targetType) query.where.targetType = targetType;
        if (targetId) query.where.targetId = targetId;

        // Date range filter
        if (startDate || endDate) {
            query.where.timestamp = {};
            if (startDate) query.where.timestamp.$gte = new Date(startDate);
            if (endDate) query.where.timestamp.$lte = new Date(endDate);
        }

        // Search filter
        if (search) {
            query.where.$or = [
                { description: { $like: `%${search}%` } },
                { actorEmail: { $like: `%${search}%` } },
                { targetName: { $like: `%${search}%` } },
            ];
        }

        const [results, total] = await Promise.all([
            this.db.findMany(query),
            this.db.count({ table: this.table, where: query.where }),
        ]);

        return {
            data: results.map(r => new AuditLog(r)),
            pagination: {
                page,
                perPage,
                total,
                totalPages: Math.ceil(total / perPage),
            },
        };
    }

    /**
     * Get logs for a specific actor
     *
     * @param {string} actorId - Actor ID
     * @param {Object} options - Query options
     * @returns {Promise<Object>}
     */
    async findByActor(actorId, options = {}) {
        return this.query({ ...options, actorId });
    }

    /**
     * Get logs for a specific target
     *
     * @param {string} targetType - Target type
     * @param {string} targetId - Target ID
     * @param {Object} options - Query options
     * @returns {Promise<Object>}
     */
    async findByTarget(targetType, targetId, options = {}) {
        return this.query({ ...options, targetType, targetId });
    }

    /**
     * Get security events
     *
     * @param {Object} options - Query options
     * @returns {Promise<Object>}
     */
    async getSecurityEvents(options = {}) {
        const securityCategories = ['auth', 'access'];

        return this.query({
            ...options,
            where: {
                $or: [
                    { eventCategory: { $in: securityCategories } },
                    { severity: { $in: ['warning', 'error', 'critical', 'alert'] } },
                ],
            },
        });
    }

    /**
     * Get failed events
     *
     * @param {Object} options - Query options
     * @returns {Promise<Object>}
     */
    async getFailedEvents(options = {}) {
        return this.query({ ...options, outcome: OUTCOME_STATUS.FAILURE });
    }

    /**
     * Count events by type in date range
     *
     * @param {Date} startDate - Start date
     * @param {Date} endDate - End date
     * @returns {Promise<Object>}
     */
    async countByType(startDate, endDate) {
        const query = {
            table: this.table,
            select: ['eventType', 'COUNT(*) as count'],
            where: {
                timestamp: {
                    $gte: startDate,
                    $lte: endDate,
                },
            },
            groupBy: ['eventType'],
        };

        return this.db.aggregate(query);
    }

    /**
     * Export logs to CSV format
     *
     * @param {Object} options - Query options
     * @returns {Promise<string>} CSV string
     */
    async exportToCsv(options = {}) {
        const result = await this.query({ ...options, perPage: 10000 });
        const logs = result.data;

        const headers = [
            'ID', 'Timestamp', 'Event Type', 'Category', 'Severity', 'Outcome',
            'Actor ID', 'Actor Email', 'Actor IP',
            'Target Type', 'Target ID', 'Target Name',
            'Description',
        ];

        const rows = logs.map(log => [
            log.id,
            log.timestamp.toISOString(),
            log.eventType,
            log.eventCategory,
            log.severity,
            log.outcome,
            log.actorId || '',
            log.actorEmail || '',
            log.actorIp || '',
            log.targetType || '',
            log.targetId || '',
            log.targetName || '',
            (log.description || '').replace(/"/g, '""'),
        ]);

        const csvRows = [
            headers.join(','),
            ...rows.map(row => row.map(cell => `"${cell}"`).join(',')),
        ];

        return csvRows.join('\n');
    }
}

// -----------------------------------------------------------------------------
// Module Exports
// -----------------------------------------------------------------------------

module.exports = {
    AuditLog,
    AuditLogBuilder,
    AuditLogSchema,
    AuditLogRepository,
    AUDIT_EVENT_TYPES,
    SEVERITY_LEVELS,
    OUTCOME_STATUS,
};

File: src/backend/app/models/audit_log.py
"""
Audit log model for compliance and security tracking.
Implements immutable audit trail with builder pattern.
"""
from datetime import datetime
from enum import Enum
from typing import Optional, Any, Dict
from uuid import uuid4

from sqlalchemy import String, Integer, Boolean, DateTime, Text, Index
from sqlalchemy.dialects.postgresql import UUID, JSONB, INET
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base


class AuditEventType(str, Enum):
    """Types of audit events."""
    # Authentication
    AUTH_LOGIN_SUCCESS = "auth.login.success"
    AUTH_LOGIN_FAILED = "auth.login.failed"
    AUTH_LOGOUT = "auth.logout"
    AUTH_MFA_ENROLLED = "auth.mfa.enrolled"
    AUTH_MFA_CHALLENGE = "auth.mfa.challenge"
    AUTH_PASSWORD_RESET = "auth.password.reset"
    AUTH_PASSWORD_CHANGED = "auth.password.changed"

    # User management
    USER_CREATED = "user.created"
    USER_UPDATED = "user.updated"
    USER_DELETED = "user.deleted"
    USER_BLOCKED = "user.blocked"
    USER_UNBLOCKED = "user.unblocked"

    # Role management
    ROLE_ASSIGNED = "role.assigned"
    ROLE_REMOVED = "role.removed"

    # Team management
    TEAM_CREATED = "team.created"
    TEAM_UPDATED = "team.updated"
    TEAM_DELETED = "team.deleted"
    TEAM_MEMBER_ADDED = "team.member.added"
    TEAM_MEMBER_REMOVED = "team.member.removed"

    # Organization
    ORG_CREATED = "org.created"
    ORG_UPDATED = "org.updated"
    ORG_DELETED = "org.deleted"

    # Access control
    ACCESS_DENIED = "access.denied"
    ACCESS_GRANTED = "access.granted"

    # Compliance
    COMPLIANCE_REPORT_GENERATED = "compliance.report.generated"
    COMPLIANCE_EXPORT = "compliance.export"

    # Admin actions
    ADMIN_OVERRIDE = "admin.override"
    ADMIN_CONFIG_CHANGED = "admin.config.changed"

    # System
    SYSTEM_ERROR = "system.error"
    SYSTEM_STARTUP = "system.startup"
    SYSTEM_SHUTDOWN = "system.shutdown"


class AuditSeverity(str, Enum):
    """Severity levels for audit events."""
    DEBUG = "debug"
    INFO = "info"
    NOTICE = "notice"
    WARNING = "warning"
    ERROR = "error"
    CRITICAL = "critical"
    ALERT = "alert"


class AuditOutcome(str, Enum):
    """Outcome of the audited action."""
    SUCCESS = "success"
    FAILURE = "failure"
    UNKNOWN = "unknown"


class AuditLog(Base):
    """
    Immutable audit log entry.
    Designed for compliance requirements (SOC 2, HIPAA, GDPR).
    """

    __tablename__ = "audit_logs"

    # Primary key
    id: Mapped[str] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid4,
    )

    # Timestamp (immutable)
    timestamp: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=datetime.utcnow,
        nullable=False,
        index=True,
    )

    # Event classification
    event_type: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
        index=True,
    )
    event_category: Mapped[Optional[str]] = mapped_column(
        String(50),
        nullable=True,
        index=True,
    )
    severity: Mapped[str] = mapped_column(
        String(20),
        default=AuditSeverity.INFO.value,
        nullable=False,
    )
    outcome: Mapped[str] = mapped_column(
        String(20),
        default=AuditOutcome.SUCCESS.value,
        nullable=False,
    )

    # Actor information
    actor_id: Mapped[Optional[str]] = mapped_column(
        UUID(as_uuid=True),
        nullable=True,
        index=True,
    )
    actor_type: Mapped[Optional[str]] = mapped_column(
        String(50),
        nullable=True,
    )
    actor_email: Mapped[Optional[str]] = mapped_column(
        String(254),
        nullable=True,
    )
    actor_ip: Mapped[Optional[str]] = mapped_column(
        INET,
        nullable=True,
    )
    actor_user_agent: Mapped[Optional[str]] = mapped_column(
        Text,
        nullable=True,
    )

    # Target information
    target_type: Mapped[Optional[str]] = mapped_column(
        String(50),
        nullable=True,
    )
    target_id: Mapped[Optional[str]] = mapped_column(
        String(255),
        nullable=True,
        index=True,
    )
    target_name: Mapped[Optional[str]] = mapped_column(
        String(255),
        nullable=True,
    )

    # Organization context
    organization_id: Mapped[Optional[str]] = mapped_column(
        UUID(as_uuid=True),
        nullable=True,
        index=True,
    )

    # Event details
    description: Mapped[Optional[str]] = mapped_column(
        Text,
        nullable=True,
    )
    changes: Mapped[Optional[dict]] = mapped_column(
        JSONB,
        nullable=True,
    )
    metadata: Mapped[Optional[dict]] = mapped_column(
        JSONB,
        nullable=True,
    )

    # Request context
    request_id: Mapped[Optional[str]] = mapped_column(
        String(100),
        nullable=True,
        index=True,
    )
    session_id: Mapped[Optional[str]] = mapped_column(
        String(100),
        nullable=True,
    )

    # Geographic information
    geo_country: Mapped[Optional[str]] = mapped_column(
        String(2),
        nullable=True,
    )
    geo_city: Mapped[Optional[str]] = mapped_column(
        String(100),
        nullable=True,
    )

    # Hash chain for tamper detection
    previous_hash: Mapped[Optional[str]] = mapped_column(
        String(64),
        nullable=True,
    )
    current_hash: Mapped[Optional[str]] = mapped_column(
        String(64),
        nullable=True,
    )

    # Indexes for common queries
    __table_args__ = (
        Index("ix_audit_org_timestamp", "organization_id", "timestamp"),
        Index("ix_audit_actor_timestamp", "actor_id", "timestamp"),
        Index("ix_audit_type_org", "event_type", "organization_id"),
    )

    def __repr__(self) -> str:
        return f"<AuditLog {self.event_type} at {self.timestamp}>"

    @property
    def is_security_event(self) -> bool:
        """Check if this is a security-related event."""
        security_types = ["auth.", "access.", "admin."]
        return any(self.event_type.startswith(t) for t in security_types)

    @property
    def is_failure(self) -> bool:
        """Check if this event represents a failure."""
        return self.outcome == AuditOutcome.FAILURE.value

    @property
    def is_high_severity(self) -> bool:
        """Check if this is a high severity event."""
        high_severities = [
            AuditSeverity.ERROR.value,
            AuditSeverity.CRITICAL.value,
            AuditSeverity.ALERT.value,
        ]
        return self.severity in high_severities


class AuditLogBuilder:
    """
    Builder pattern for creating audit log entries.
    Provides fluent interface for constructing audit logs.
    """

    def __init__(self):
        self._event_type: Optional[str] = None
        self._event_category: Optional[str] = None
        self._severity: str = AuditSeverity.INFO.value
        self._outcome: str = AuditOutcome.SUCCESS.value
        self._actor_id: Optional[str] = None
        self._actor_type: Optional[str] = None
        self._actor_email: Optional[str] = None
        self._actor_ip: Optional[str] = None
        self._actor_user_agent: Optional[str] = None
        self._target_type: Optional[str] = None
        self._target_id: Optional[str] = None
        self._target_name: Optional[str] = None
        self._organization_id: Optional[str] = None
        self._description: Optional[str] = None
        self._changes: Optional[dict] = None
        self._metadata: Dict[str, Any] = {}
        self._request_id: Optional[str] = None
        self._session_id: Optional[str] = None
        self._geo_country: Optional[str] = None
        self._geo_city: Optional[str] = None

    def event(self, event_type: str) -> "AuditLogBuilder":
        """Set event type."""
        self._event_type = event_type
        self._event_category = event_type.split(".")[0] if "." in event_type else None
        return self

    def severity(self, severity: AuditSeverity) -> "AuditLogBuilder":
        """Set severity level."""
        self._severity = severity.value if isinstance(severity, AuditSeverity) else severity
        return self

    def success(self) -> "AuditLogBuilder":
        """Mark as successful outcome."""
        self._outcome = AuditOutcome.SUCCESS.value
        return self

    def failure(self) -> "AuditLogBuilder":
        """Mark as failed outcome."""
        self._outcome = AuditOutcome.FAILURE.value
        return self

    def actor(
        self,
        actor_id: str = None,
        actor_type: str = None,
        email: str = None,
        ip: str = None,
        user_agent: str = None,
    ) -> "AuditLogBuilder":
        """Set actor information."""
        self._actor_id = actor_id
        self._actor_type = actor_type
        self._actor_email = email
        self._actor_ip = ip
        self._actor_user_agent = user_agent
        return self

    def target(
        self,
        target_type: str,
        target_id: str = None,
        target_name: str = None,
    ) -> "AuditLogBuilder":
        """Set target information."""
        self._target_type = target_type
        self._target_id = target_id
        self._target_name = target_name
        return self

    def organization(self, org_id: str) -> "AuditLogBuilder":
        """Set organization context."""
        self._organization_id = org_id
        return self

    def describe(self, description: str) -> "AuditLogBuilder":
        """Set description."""
        self._description = description
        return self

    def changes(self, before: dict = None, after: dict = None) -> "AuditLogBuilder":
        """Set changes (before/after)."""
        self._changes = {"before": before, "after": after}
        return self

    def meta(self, **kwargs) -> "AuditLogBuilder":
        """Add metadata."""
        self._metadata.update(kwargs)
        return self

    def request(
        self,
        request_id: str = None,
        session_id: str = None,
    ) -> "AuditLogBuilder":
        """Set request context."""
        self._request_id = request_id
        self._session_id = session_id
        return self

    def geo(self, country: str = None, city: str = None) -> "AuditLogBuilder":
        """Set geographic information."""
        self._geo_country = country
        self._geo_city = city
        return self

    def build(self) -> AuditLog:
        """Build the AuditLog instance."""
        if not self._event_type:
            raise ValueError("Event type is required")

        return AuditLog(
            event_type=self._event_type,
            event_category=self._event_category,
            severity=self._severity,
            outcome=self._outcome,
            actor_id=self._actor_id,
            actor_type=self._actor_type,
            actor_email=self._actor_email,
            actor_ip=self._actor_ip,
            actor_user_agent=self._actor_user_agent,
            target_type=self._target_type,
            target_id=self._target_id,
            target_name=self._target_name,
            organization_id=self._organization_id,
            description=self._description,
            changes=self._changes,
            metadata=self._metadata if self._metadata else None,
            request_id=self._request_id,
            session_id=self._session_id,
            geo_country=self._geo_country,
            geo_city=self._geo_city,
        )
